Makefile.one

```diff
@@ -31,24 +32,73 @@ CMD_CAT ?= cat
 CMD_MD5 ?= md5sum
 CMD_OPA ?= opa # https://github.com/open-policy-agent/opa/releases/download/v0.33.1/opa_linux_amd64
 
-check_%:
-	@command -v $* >/dev/null || (echo "missing required tool $*" ; exit 1)
+.check_%:
+#
+	@command -v $* >/dev/null
+	if [ $$? -ne 0 ]; then
+		echo "missing required tool $*"
+		exit 1
+	else
+		touch $@ # avoid target rebuilds due to inexistent file
+	fi
 
+#
 # libs
+#
 
 LIB_ELF ?= libelf
 LIB_ZLIB ?= zlib
 
-checklib_%:
-	@$(CMD_PKGCONFIG) --silence-errors --validate $* 2>/dev/null || \
-		(echo "missing lib $*" ; exit 1)
+.checklib_%: \
+	| .check_$(CMD_PKGCONFIG)
+#
+	@$(CMD_PKGCONFIG) --silence-errors --validate $* 2>/dev/null
+	if [ $$? -ne 0 ]; then
+	        echo "missing lib $*"
+		exit 1
+	else
+		touch $@ # avoid target rebuilds due to inexistent file
+	fi
+
+#
+# tools version
+#
+
+CLANG_VERSION = $(shell $(CMD_CLANG) --version 2>/dev/null | \
+	head -1 | tr -d '[:alpha:]' | tr -d '[:space:]' | cut -d'.' -f1)
 
+.checkver_$(CMD_CLANG): \
+	| .check_$(CMD_CLANG)
+#
+	@if [ ${CLANG_VERSION} -lt 12 ]; then
+		echo -n "make sure you're using clang 12 or higher, "
+		echo "your current clang version is ${CLANG_VERSION}"
+		exit 1
+	fi
+	touch $@ # avoid target rebuilds over and over due to inexistent file
+
+GO_VERSION = $(shell $(CMD_GO) version 2>/dev/null | awk '{print $$3}' | sed 's:go::g' | cut -d. -f1,2)
+
+.checkver_$(CMD_GO): \
+	| .check_$(CMD_GO)
+#
+	@if [ "${GO_VERSION}" != "1.16" ] && [ "${GO_VERSION}" != "1.17" ]; then
+		echo -n "you MUST use golang 1.16 or 1.17, "
```

Comment:
This will break when a newer version of go will be out.
Maybe we should check if version is higher or equal to 1.16 instead?

Reply:
I thought about it and decided to bind it to 1.16 or 1.17 because we know, for sure, they are compatible. We don't know if our code is compatible to 1.18 (or 1.19, ...) and the idea was to hard code the supported versions here. With that said, do you still want to make that change ?

Reply:
So you think that go is not backwards compatible? I'm not sure this is true, but I'm also not 100% sure about that

Reply:
@mtcherni95 any opinion here? Should we do specific versioning tests ? Or assume anything higher than 1.16 would be compatible ?

Reply:
I agree with Yaniv, that we need to check for minimum version (if we use a new language feature not available previously), but not for maximum version (if existing code will stop working), since Go is supposed to be forwards compatible, and in rare cases where it's not it would be a matter of bug fix